packages/opentelemetry-plugin-express/src/express.ts

```diff
@@ -0,0 +1,215 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// mongodb.Server type is deprecated so every use trigger a lint error
+/* tslint:disable:deprecation */
+
+import { BasePlugin } from '@opentelemetry/core';
+import { CanonicalCode, Span, Attributes } from '@opentelemetry/types';
+import * as express from 'express';
+import * as shimmer from 'shimmer';
+import {
+  ExpressLayer,
+  ExpressRouter,
+  AttributeNames,
+  PatchedRequest,
+} from './types';
+import { VERSION } from './version';
+
+export const kPatched: unique symbol = Symbol('express-layer-patched');
+
+/** Express instrumentation plugin for OpenTelemetry */
+export class ExpressPlugin extends BasePlugin<typeof express> {
+  private readonly _COMPONENT = 'express';
+
+  readonly supportedVersions = ['^4.0.0'];
+
+  constructor(readonly moduleName: string) {
+    super('@opentelemetry/plugin-express', VERSION);
+  }
+
+  /**
+   * Patches Express operations.
+   */
+  protected patch() {
+    this._logger.debug('Patching Express');
+
+    if (this._moduleExports) {
+      const routerProto = (this._moduleExports
+        .Router as unknown) as express.Router;
+      const plugin = this;
+
+      this._logger.debug('patching express.Router.prototype.route');
+      shimmer.wrap(routerProto, 'route', (original: Function) => {
+        return function route_trace(
+          this: ExpressRouter,
+          arg: string | Function
+        ) {
+          const route = original.apply(this, arguments);
+          const layer = this.stack[this.stack.length - 1] as ExpressLayer;
+          plugin._applyPatch(layer, typeof arg === 'string' ? arg : undefined);
+          return route;
+          // tslint:disable-next-line:no-any
+        } as any;
+      });
+      this._logger.debug('patching express.Router.prototype.use');
+      shimmer.wrap(routerProto, 'use', (original: Function) => {
+        return function use(this: express.Application, arg: string | Function) {
+          const route = original.apply(this, arguments);
+          const layer = this.stack[this.stack.length - 1] as ExpressLayer;
+          plugin._applyPatch(layer, typeof arg === 'string' ? arg : undefined);
```

Comment:
It's simply because you can mount a layer without a path, that's the case for all middlewares:
```
express.use(function (req, res, next) {
 // code
})
```